### PR TITLE
Fixed issue where DRBG_CTR fails if NO_DF is used

### PR DESCRIPTION
**Pull request by slontis**
When rand_drbg_get_entropy() is called with the CTR DRBG and the NO_DF flag is used - the maximum size of the additional length passed to the generate() is the size of the seed_len (e.g- 48 bytes). The sizeof(*drbg) is hundreds of bytes. This has not been picked up before because most of the tests (related to CAVS) dont use real entropy. Also the way the set_defaults() works, the master DRBG is only initialized once - which is why the NO_DF case has not really been tested. with real entropy. I have tested this after making other changes to the DRBG code (which is in a seperate pull request).

### CODE REVIEW

**Comment by t-j-h** [comment]
It seems conceptually to be fixing this issue at the wrong location. The max_adinlen is one of those items noted as pretty much optionally set up by the per-type init function. Now we are checking those values up in the level above.

I would have expected other per-type classes to not actually set those variables - i.e. they are there to be optionally used - and there may actually not be such a concept - a maximum length in this context in some per-type implementation.

At the very least the comments in rand_lcl.h about these variables and needing to always be set (for at least this variable) needs to be made clear.

**Comment by kroeckx** [comment]
The other solution is not to pass the drbg as additional data. If we want additional data there we can get it using rand_drbg_get_additional_data().

**Comment by slontis** [comment]
Without the change the current code will **ALWAYS** fail if CTR with NO_DF is used.
The values described in rand_lcl.h just refer to the tables in SP800-90Ar1.
The DF case is no problem since it is very large, its only the NO_DF case which has a max of
seedlen which is a problem.

I dont see a nice way of moving in down..
Currently In RAND_DRBG_generate() it  does
  if (adinlen > drbg->max_adinlen) { error } 
before it calls
drbg->meth->generate().

You could truncate the value instead at this layer instead of doing a error, but I dont know that this is a good idea. Shifting the check into the meth->generate() is also messy as it would need to be done for each type.

**Comment by kroeckx** [comment]
This is a little bit unrelated to this PR, but the NO_DF version is not FIPS compliant as far as we understand because you need to get full entropy from the parent DRBG, which currently can't give that.

**Comment by slontis** [comment]
Agreed. But that is only because openssl doesnt supply FIPS compliant entropy sources.
There is CAVS test data for the no_df case.

File: crypto/rand/rand_lib.c [hunk 1]
```diff
@@ -566,8 +566,7 @@ size_t rand_pool_bytes_needed(RAND_POOL *pool, unsigned int entropy_factor)
         return 0;
     }
 
-    if (pool->len < pool->min_len &&
-        bytes_needed < pool->min_len - pool->len)
+    if (pool->len < pool->min_len && bytes_needed < pool->min_len - pool->len)
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: I don't see any necessity to make this unrelated whitespace-only change. It only distracts from the real change. With GitHub fine-diffs it's no problem to see that it's whitespace-only, but otherwise one has to look twice.

File: crypto/rand/rand_lib.c [hunk 2]
```diff
@@ -174,7 +174,7 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
             if (RAND_DRBG_generate(drbg->parent,
                                    buffer, bytes_needed,
                                    prediction_resistance,
-                                   (unsigned char *)drbg, sizeof(*drbg)) != 0)
+                                   NULL, 0) != 0)
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: For the sake of simplicity, this is probably the best one can do here.

**Comment by paulidale** [comment]
The white space change can be reverted, but it doesn't seem harmful with the revised commit which is very simple).

File: crypto/rand/rand_lib.c (hunk 2 above)
Comment (slontis) [reply crypto/rand/rand_lib.c]: Done

Comment (mspncp) [reply crypto/rand/rand_lib.c]: Thanks!

**Comment by paulidale** [comment]
Merged, thanks to all.

**Comment by kaduk** [comment]
I'm failing to remember if adding the drbg contents as AD was part of a scheme to ensure disambiguation between sibling DRBGs.  Though presumably the entropy from the parent would be enough to do that...

**Comment by paulidale** [comment]
The seeding from the parent is sufficient for this.  The AD, in this case, is a function of the previous seeding.
